### PR TITLE
Error handling for adding a relay to a relay group

### PR DESCRIPTION
**Pull request by lhaskins**
Errors were not being passed correctly when there was a failure when adding a nonexistent relay to a relay group.

Fixes https://github.com/operable/cog/issues/535

### CODE REVIEW

File: lib/cog_api/http/relay_groups.ex [hunk 1]
```diff
@@ -39,16 +39,24 @@ defmodule CogApi.HTTP.RelayGroups do
   end
 
   def add_relay(%{name: name}, %{relay: relay_name}, %Endpoint{}=endpoint) do
```
Comment (mpeck) [inline lib/cog_api/http/relay_groups.ex]: So I ended up refactoring the add_bundle bit in my branch. The reason I mention it is because it is super similar to this bit. The `add_relay` and `remove_relay` functions actually use the exact same logic as one another. The only difference is when they pass off to `update_memberships` they pass an atom to either `:add` or `:remove` the relay. Like I said, this is super similar to `add_bundle` and `remove_bundle`. I ended up refactoring and just merging both functions into `update_assignments`. Curious what your thoughts are on that? https://github.com/operable/cog-api-client/pull/89/files#diff-e6847729ab31281c5ad0f078cc6a8f23R70

**Comment by mpeck** [comment]
Looks good to me :+1:

File: lib/cog_api/http/relay_groups.ex [hunk 2]
```diff
@@ -38,27 +38,23 @@ defmodule CogApi.HTTP.RelayGroups do
...
-    update_membership(relay_group.id, relay.id, :add, endpoint)
-  end
-  def add_relay(relay_group_id, relay_id, %Endpoint{}=endpoint) do
-    update_membership(relay_group_id, relay_id, :add, endpoint)
-  end
-
-  def remove_relay(%{name: name}, %{relay: relay_name}, %Endpoint{}=endpoint) do
-    {relay, relay_group} = get_group_relay(name, relay_name, endpoint)
-    update_membership(relay_group.id, relay.id, :remove, endpoint)
+  def update_memberships(action, %{name: name}, %{relay: relay_name}, %Endpoint{}=endpoint) do
```
Comment (mpeck) [inline lib/cog_api/http/relay_groups.ex]: Can we make `update_memberships` take a list of relays? That will correspond to the pattern in `update_assignments`. The api already supports it.

File: lib/cog_api/fake/relay_groups.ex [hunk 3]
```diff
@@ -52,13 +52,18 @@ defmodule CogApi.Fake.RelayGroups do
...
     relay_group = Server.show_by_key(RelayGroup, :name, name)
-    add_relay(relay, relay_group)
+    relay_ids = Enum.map(relay_names, &Server.show_by_key(Relay, :name, &1))
+    |> Enum.map(&Map.fetch!(&1, :id))
+    add_relays(relay_group.id, relay_ids, endpoint)
   end
-  def add_relay(id, relay_id, %Endpoint{token: _}) do
+  def add_relays(id, relay_ids, %Endpoint{token: _}=endpoint) when is_list(relay_ids) do
+    Enum.map(relay_ids, &add_relays(id, &1, endpoint))
+    |> List.last
```
Comment (christophermaier) [inline lib/cog_api/fake/relay_groups.ex]: Why `List.last`?

File: lib/cog_api/http/relay_groups.ex [hunk 4]
```diff
@@ -38,30 +38,19 @@ defmodule CogApi.HTTP.RelayGroups do
     |> ApiResponse.format_delete("The relay group could not be deleted")
   end
 
-  def update_memberships(action, %{name: name}, %{relay: relay_name}, %Endpoint{}=endpoint) do
-    case get_group_relay(name, relay_name, endpoint) do
-      {:ok, relay, relay_group} ->
-        update_membership(relay_group.id, relay.id, action, endpoint)
-      error ->
-        error
+  def update_memberships(action, %{name: name}, %{relays: relay_names}, endpoint) when is_list(relay_names) do
```
Comment (christophermaier) [inline lib/cog_api/http/relay_groups.ex]: Is this function ever publicly called with anything other than names? The maps-as-a-differentiator here make for confusing-to-read code.

File: lib/cog_api/http/relay_groups.ex [hunk 5]
```diff
@@ -38,30 +38,19 @@ defmodule CogApi.HTTP.RelayGroups do
...
-  end
-
-  defp update_membership(relay_group_id, relay_id, action, endpoint) do
+  def update_memberships(action, relay_group_id, relay_ids, endpoint) when is_list(relay_ids) do
     path = "relay_groups/#{relay_group_id}/relays"
-    Base.post(endpoint, path, %{relays: %{action =>  [relay_id]}})
+    Base.post(endpoint, path, %{relays: %{action => relay_ids}})
     |> ApiResponse.format(%{"relay_group" => RelayGroup.format})
   end
+  def update_memberships(action, relay_group_id, relay_id, endpoint) when is_binary(relay_id),
```
Comment (christophermaier) [inline lib/cog_api/http/relay_groups.ex]: You can also eliminate this function head and the guards (as well as any others that are similar) by simply wrapping `relay_ids` with `List.wrap` before you do anything else.

In that case, I'd also rename `relay_ids` to `relay_id_or_ids` to be a little more clear, but only if calling with a single ID is still actually something that can occur.

File: lib/cog_api/http/relay_groups.ex (hunk 4 above)
Comment (mpeck) [inline lib/cog_api/http/relay_groups.ex]: Yeah, I agree about the maps. I'm taking this as an iterative refactor. The function is called publicly with either a map or an id. Both update_memberships and update_assignments use the same pattern. I'm thinking that we can clean them up and consolidate the code a bit, but like I said, I was saving that for a later refactor.

File: lib/cog_api/http/relay_groups.ex [hunk 6]
```diff
@@ -77,6 +66,26 @@ defmodule CogApi.HTTP.RelayGroups do
...
+    end
+
+    case Enum.reduce_while(relay_names, [], &get_relay_id.(&1, &2, endpoint)) do
+      relay_ids when is_list(relay_ids) ->
+        {:ok, relay_ids}
+      error ->
+        error
+    end
+  end
+  defp get_relay_ids(relay_name, endpoint),
```
Comment (christophermaier) [inline lib/cog_api/http/relay_groups.ex]: I'd probably also use `List.wrap` instead of an extra function head here, too

File: lib/cog_api/fake/relay_groups.ex (hunk 3 above)
Comment (mpeck) [inline lib/cog_api/fake/relay_groups.ex]: I don't remember. I'm pretty sure there was a reason, I'm just not sure if it was a good one or not.

File: lib/cog_api/http/relay_groups.ex [hunk 7]
```diff
@@ -77,6 +66,26 @@ defmodule CogApi.HTTP.RelayGroups do
   def update_assignments(action, relay_group_id, bundle_id, endpoint) when is_binary(bundle_id),
     do: update_assignments(action, relay_group_id, [bundle_id], endpoint)
 
+  defp get_relay_ids(relay_names, endpoint) when is_list(relay_names) do
```
Comment (christophermaier) [inline lib/cog_api/http/relay_groups.ex]: It'd be nice to abstract this "turn names into IDs" behavior into its own function, since it's the same for relays and bundles, particularly if this same pattern is also being used elsewhere, which I suspect it probably is.

File: test/fixtures/vcr/relay_group_add_relay.json [hunk 8]
```diff
@@ -11,14 +11,14 @@
       "url": "http://localhost:4000/v1/token"
     },
     "response": {
-      "body": "{\"token\":{\"value\":\"w05n5iZL4peQoZ1gibP1MR1TgvSTXrIwZzCIJYRjR+0=\"}}",
```
Comment (christophermaier) [inline test/fixtures/vcr/relay_group_add_relay.json]: It doesn't look like anything significant actually changed in this fixture, and I think a few others are like that, too. If that is indeed the case, it'd be nice to exclude these changes from the PR, just to keep it lean and focused.

File: lib/cog_api/fake/relay_groups.ex (hunk 3 above)
Comment (christophermaier) [inline lib/cog_api/fake/relay_groups.ex]: Sounds suspicious...

File: lib/cog_api/http/relay_groups.ex (hunk 4 above)
Comment (christophermaier) [inline lib/cog_api/http/relay_groups.ex]: If it's a big refactor, then maybe we can wait a bit. If it's easy, though, it'd be nice to nip it in the bud now. Sometimes these patterns tend to metastasize if they stay around too long.

Comment (mpeck) [inline lib/cog_api/http/relay_groups.ex]: @christophermaier I think the major issue actually revolves around how the function is being called. It's dependent on that map with the name key being there. cogctl calls these functions so it's a bit more far reaching. If I can I would like to go ahead and get this merged, but I can see about refactoring next. If that sounds reasonable?

File: lib/cog_api/http/relay_groups.ex (hunk 6 above)
Comment (mpeck) [inline lib/cog_api/http/relay_groups.ex]: Cool, I did not know about List.wrap

File: test/fixtures/vcr/relay_group_add_relay.json (hunk 8 above)
Comment (mpeck) [inline test/fixtures/vcr/relay_group_add_relay.json]: What do you mean? VCR forces you to rerecord the cassette if you change the test. Or am I misunderstanding?

Comment (mpeck) [inline test/fixtures/vcr/relay_group_add_relay.json]: I'm not sure I understand @christophermaier. I'm with you on keeping the PR focused, but how do I exclude the vcr bits and have the tests pass? I thought you had to rerecord the cassette whenever the test changed.

File: lib/cog_api/http/client.ex [hunk 9]
```diff
@@ -134,20 +134,36 @@ defmodule CogApi.HTTP.Client do
     RelayGroups.delete(relay_group_id, endpoint)
   end
 
-  def relay_group_add_relay(relay_group_id, relay_id, %Endpoint{}=endpoint) do
-    RelayGroups.update_memberships(:add, relay_group_id, relay_id, endpoint)
+  def relay_group_add_relays_by_name(relay_group_name, relay_names, %Endpoint{}=endpoint) do
```
Comment (mpeck) [inline lib/cog_api/http/client.ex]: @christophermaier what do you think about this pattern. It seems a lot clearer to me.

**Comment by mpeck** [comment]
@christophermaier so this has turned into a bigger refactor than was originally intended, but I think it's for the better. I split the `update_memberships` and `update_assignments` into multiple functions, `update_memberships_by_name`, `update_memberships_by_id`, `update_assignments_by_name` and `update_assignements_by_id`. I think that's a lot less confusing than the whole matching on an arbitrary map thing.

A couple things to note: This will require a few change in cogctl since it calls some of the renamed functions. And also, the matching on the map thing is a pattern that I've seen in several places. Maybe this is the beginnings of a bigger refactor?

Anyway, let me know your thoughts on this. If need be I can always roll back to the way it was before.

File: lib/cog_api/http/client.ex (hunk 9 above)
Comment (lhaskins) [inline lib/cog_api/http/client.ex]: So while have the function name denotes the expected inputs and that seems pretty clear what should be expected, I'm not sure that this is the pattern that we want to move forward with. Is there a problem with having a single function with multiple function heads that can take a list of ids or a list of names? This way we can call `relay_group_add_relays` and depending on the data sent, we can act accordingly.

File: test/fixtures/vcr/relay_group_add_relay.json (hunk 8 above)
Comment (christophermaier) [inline test/fixtures/vcr/relay_group_add_relay.json]: I mean it looks like the only things that have changed in this cassette are things like tokens, timestamps, and request IDs, all of which are incidental when we're talking about testing requests and responses. There are other cassettes where there are actual meaningful changes to things, and those should of course be kept.

It's not a huge deal, but if these changes aren't necessary, then why keep them?

File: lib/cog_api/fake/client.ex [hunk 10]
```diff
@@ -134,20 +134,36 @@ defmodule CogApi.Fake.Client do
...
-    RelayGroups.remove_bundles(relay_group_id, bundle_id, endpoint)
+  def relay_group_remove_relays_by_id(relay_group_id, relay_ids, %Endpoint{}=endpoint) do
+    RelayGroups.remove_relays_by_id(relay_group_id, relay_ids, endpoint)
+  end
+
+  def relay_group_add_bundles_by_name(relay_group_name, bundle_names, %Endpoint{}=endpoint) do
+    RelayGroups.add_bundles_by_name(relay_group_name, bundle_names, endpoint)
+  end
+
+  def relay_group_add_bundles_by_id(relay_group_id, bundle_ids, %Endpoint{}=endpoint) do
```
Comment (christophermaier) [inline lib/cog_api/fake/client.ex]: Are these `by_id` functions actually used publicly? If not, they can just be internal details, and we get a simpler API

File: lib/cog_api/http/client.ex (hunk 9 above)
Comment (mpeck) [inline lib/cog_api/http/client.ex]: @lhaskins there are a few reasons:

* One is that you have to know more about how the function is written in order to call it. So for example you have to know that you can either call the function passing a relay group id and a list of relay/bundle ids or if you want to pass relay group and relay/bundle names you need to create a map and insert the values in the proper field. The only reason for the map is to match the proper function head, it really has no bearing on the data itself.
* It also makes reasoning about the code easier. You can quickly look at the function definition and understand both what the expected data should look like and how that data will flow through the function.
* It just makes for a more clear statement of intent when you are calling the function. It's easy to look at the code and determine what it's supposed to do.

File: lib/cog_api/fake/client.ex (hunk 10 above)
Comment (mpeck) [inline lib/cog_api/fake/client.ex]: I honestly don't know @christophermaier. The original implementation exposed the ability to call the functions with either an id or a name so that's why I included both. `cogctl` only uses the `_by_name` version, but Flywheel may be using the `_by_id` bits.

File: test/fixtures/vcr/relay_group_add_relay.json (hunk 8 above)
Comment (mpeck) [inline test/fixtures/vcr/relay_group_add_relay.json]: @christophermaier Oh, I'm not arguing that unnecessary changes shouldn't be dropped. When I wrote the tests vcr complained and would not run until I regenerated the cassettes. I'm just saying that I don't know how to prevent that from happening.

File: lib/cog_api/fake/client.ex (hunk 10 above)
Comment (mpeck) [inline lib/cog_api/fake/client.ex]: They are used in the tests.

Comment (christophermaier) [inline lib/cog_api/fake/client.ex]: That'd be a good thing to check re: Flywheel, then.
